File: reqy-web/scripts/refactor-layout.py
#!/usr/bin/env python3
"""One-shot refactor: move sidebar/header/wrapper out of each page so (app)/layout.tsx owns them."""

import os
import re
import sys

ROOT = os.path.join(os.getcwd(), "app", "(app)")
PAGES = [
    "page.tsx",
    "dashboard/page.tsx",
    "collections/page.tsx",
    "mocks/page.tsx",
    "settings/page.tsx",
    "runner/page.tsx",
    "ai-insights/page.tsx",
    "documentation/page.tsx",
    "workspaces/page.tsx",
    "graphql/page.tsx",
    "my-projects/page.tsx",
    "sdks/page.tsx",
]

# Matches the opening of the wrapper: outer div + ApiSidebar + inner div + ApiHeader.
# Captures everything from the outer div opening through ApiHeader so we can drop it.
WRAPPER_OPEN = re.compile(
    r'    <div className="flex (?:h-screen|min-h-screen) bg-background(?: bg-dot-pattern)?">\s*\n'
    r'\s*<ApiSidebar[^>]*?/>\s*\n'
    r'\s*\n?'
    r'\s*<div className=\{cn\(\s*\n'
    r'[\s\S]*?'
    r'\)\}>\s*\n'
    r'\s*<ApiHeader\s*/>\s*\n'
    r'\s*\n?'
)

# Matches the closing of the wrapper: inner div close + outer div close + return close.
WRAPPER_CLOSE = re.compile(
    r'\n\s*</div>\s*\n\s*</div>\s*\n\s*\)\s*\n\s*\}\s*$'
)


def refactor(path):
    with open(path, "r", encoding="utf-8") as f:
        original = f.read()
    content = original

    # 1. Remove unused imports
    content = re.sub(r'import \{ ApiSidebar \} from "@/components/api-sidebar"\n', "", content)
    content = re.sub(r'import \{ ApiHeader \} from "@/components/api-header"\n', "", content)
    content = re.sub(r'import \{ useSidebar \} from "@/contexts/sidebar-context"\n', "", content)

    # 2. Remove useSidebar hook if its destructured names aren't used elsewhere
    hook_re = re.compile(r'  const \{ isCollapsed, toggleSidebar \} = useSidebar\(\)\n')
    if hook_re.search(content):
        candidate = hook_re.sub("", content)
        if "isCollapsed" not in candidate and "toggleSidebar" not in candidate:
            content = candidate

    # 3. Replace wrapper opening
    content = WRAPPER_OPEN.sub("", content)

    # 4. Replace wrapper closing: only match at the END of the last return statement.
    content = WRAPPER_CLOSE.sub("\n)", content)

    if content == original:
        return False, 0

    before_lines = original.count("\n")
    after_lines = content.count("\n")
    with open(path, "w", encoding="utf-8") as f:
        f.write(content)
    return True, before_lines - after_lines


def main():
    # Script lives at reqy-web/scripts/refactor-layout.py.
    # Use sys.argv[0] (always absolute when invoked with an absolute path)
    # then walk up to reqy-web/.
    script_path = os.path.realpath(sys.argv[0])
    script_dir = os.path.dirname(script_path)
    reqy_web_dir = os.path.dirname(script_dir)
    root = os.path.join(reqy_web_dir, "app", "(app)")
    changed = 0
    for rel in PAGES:
        path = os.path.join(root, rel)
        if not os.path.exists(path):
            print(f"SKIP (not found): {path}")
            continue
        ok, diff = refactor(path)
        if ok:
            changed += 1
            print(f"OK   {rel} (-{diff} lines)")
        else:
            print(f"NOOP {rel}")
    print(f"\n{changed}/{len(PAGES)} pages refactored")


if __name__ == "__main__":
    main()

File: reqy-web/scripts/refactor-layout.mjs
#!/usr/bin/env node
// One-shot refactor: move sidebar/header/wrapper out of each page so the
// (app)/layout.tsx owns them. Idempotent — running twice is a no-op.

import fs from 'node:fs'
import path from 'node:path'

const ROOT = path.resolve('app/(app)')
const PAGES = [
  'page.tsx',
  'dashboard/page.tsx',
  'collections/page.tsx',
  'mocks/page.tsx',
  'settings/page.tsx',
  'runner/page.tsx',
  'ai-insights/page.tsx',
  'documentation/page.tsx',
  'workspaces/page.tsx',
  'graphql/page.tsx',
  'my-projects/page.tsx',
  'sdks/page.tsx',
]

// Match the opening of the wrapper: outer div + ApiSidebar + inner div + ApiHeader
// Returns the "before" and "after" text.
const WRAPPER_OPEN_REGEX = /    <div className="flex (?:h-screen|min-h-screen) bg-background(?: bg-dot-pattern)?">\s*\n\s*<ApiSidebar[^>]*\/>\s*\n\s*\n?\s*<div className=\{cn\(\s*\n[\s\S]*?\}\s*\)>\s*\n\s*<ApiHeader\s*\/>\s*\n\s*\n?\s*/

// Match the closing of the wrapper: inner div close + outer div close + return close
const WRAPPER_CLOSE_REGEX = /\n\s*<\/div>\s*\n\s*<\/div>\s*\n\s*\)\s*\n\}\s*$/

function refactor(filePath) {
  let content = fs.readFileSync(filePath, 'utf8')

  const before = content

  // 1. Remove unused imports
  content = content.replace(/import \{ ApiSidebar \} from "@\/components\/api-sidebar"\n/g, '')
  content = content.replace(/import \{ ApiHeader \} from "@\/components\/api-header"\n/g, '')
  content = content.replace(/import \{ useSidebar \} from "@\/contexts\/sidebar-context"\n/g, '')

  // 2. Remove useSidebar hook (only if isCollapsed/toggleSidebar aren't used elsewhere)
  //    Heuristic: count occurrences of isCollapsed and toggleSidebar AFTER removing the destructuring.
  const hookRegex = /  const \{ isCollapsed, toggleSidebar \} = useSidebar\(\)\n/
  if (hookRegex.test(content)) {
    const withoutHook = content.replace(hookRegex, '')
    // If isCollapsed / toggleSidebar still appear elsewhere (rare), keep the hook.
    const stillUsed = /\bisCollapsed\b/.test(withoutHook) || /\btoggleSidebar\b/.test(withoutHook)
    if (!stillUsed) {
      content = withoutHook
    }
  }

  // 3. Replace wrapper opening: collapse the outer/inner div + sidebar + header
  //    into nothing (children render directly).
  const openMatch = content.match(WRAPPER_OPEN_REGEX)
  if (openMatch) {
    content = content.replace(WRAPPER_OPEN_REGEX, '')
  }

  // 4. Replace wrapper closing: the last </div></div>) in the return statement.
  //    Strategy: find the last `return (` and apply close regex only to its tail.
  const lastReturn = content.lastIndexOf('return (')
  if (lastReturn !== -1) {
    const afterReturn = content.slice(lastReturn)
    const closeMatch = afterReturn.match(WRAPPER_CLOSE_REGEX)
    if (closeMatch) {
      const closeIdx = lastReturn + afterReturn.indexOf(closeMatch[0])
      content = content.slice(0, closeIdx) + '\n  )' + content.slice(closeIdx + closeMatch[0].length)
    }
  }

  if (content === before) {
    return { changed: false }
  }

  fs.writeFileSync(filePath, content, 'utf8')
  return { changed: true, removedLines: before.split('\n').length - content.split('\n').length }
}

let changedCount = 0
for (const rel of PAGES) {
  const filePath = path.join(ROOT, rel)
  if (!fs.existsSync(filePath)) {
    console.warn(`SKIP (not found): ${filePath}`)
    continue
  }
  const result = refactor(filePath)
  if (result.changed) {
    changedCount++
    console.log(`OK  ${rel} (-${result.removedLines} lines)`)
  } else {
    console.log(`SKIP (no changes): ${rel}`)
  }
}

console.log(`\n${changedCount}/${PAGES.length} pages refactored`)
